initial sidekick implementation

Hanspagh opened the pull request:
## Sidekick implementation

I took most of the code from the prototype, but have changed the way we start the supervisor on the sidekick node. Instead of passing it as an argument to `erl`. We start it separately after the node has come up. (Added a todo because at the moment it seems even after we get the `nodeup`, the node might not be fully ready)

Added a parent supervisor instead of the naive docker implementation and moved the `parent_node` monitoring to this.

Still a bit new to using `specs`, so there might be some improvements needed there.
Let me know what you think

Hanspagh committed:
initial sidekick implementation

sasa1977 commented on lib/sidekick/sidekick.ex:
```diff
@@ -0,0 +1,81 @@
+defmodule Sidekick do
+  @spec start(atom, [{atom, any}]) :: {:error, any} | {:ok, atom, pid}
+  def start(node_name \\ :docker, children \\ [{Ci.Docker, []}]) do
```
What is `{Ci.Docker, []}`?

Hanspagh replied:
My bad that is my local docker implementation, will remove

sasa1977 commented on lib/sidekick/sidekick.ex:
```diff
@@ -0,0 +1,81 @@
+  @spec call(atom, atom, atom, list) :: any
+  def call(name \\ :docker, module, function, args) do
+    :rpc.block_call(node_host_name(name), module, function, args)
+  end
```
Let's keep this as a private function for the moment, and open it up only if we need it. Or more generally, the idea is to keep the module's API as small as possible.

sasa1977 commented on lib/sidekick/sidekick.ex:
```diff
@@ -0,0 +1,81 @@
+
+  @spec call(atom, atom, atom, list) :: any
+  def call(name \\ :docker, module, function, args) do
+    :rpc.block_call(node_host_name(name), module, function, args)
```
This seems a bit too strong. Why not `:rpc.call`?

Hanspagh replied:
On 2. note. When I change this to call, the Supervisor terminates instantly after it has been started

sasa1977 commented on lib/sidekick/sidekick.ex:
```diff
@@ -0,0 +1,81 @@
+  end
+
+  @spec start_sidekick([node]) :: :ok
+  def start_sidekick([parent_node]) do
```
This function isn't really a part of the API, but it needs to be public because of the `-s` arg. I propose adding `@doc false`.

sasa1977 commented on lib/sidekick/sidekick.ex:
```diff
@@ -0,0 +1,81 @@
+
+    receive do
+      {:nodeup, ^sidekick_node} ->
+        # wait for node to really be up
```
What does it mean to be really up?

Hanspagh replied:
If I don't add this wait, I will sometimes get an error from the rpc.call `{:badrpc, :nodedown}`

sasa1977 commented on lib/sidekick/sidekick.ex:
```diff
@@ -0,0 +1,81 @@
+      {:nodeup, ^sidekick_node} ->
+        # wait for node to really be up
+        # TODO deal with this in a better way
+        :timer.sleep(500)
```
```suggestion
        Process.sleep(500)
```

sasa1977 commented on lib/sidekick/sidekick.ex:
```diff
@@ -0,0 +1,81 @@
+        # TODO deal with this in a better way
+        :timer.sleep(500)
+
+        case call(:docker, Sidekick.Supervisor, :start_link, [[parent_node, children]]) do
```
How about dropping outer brackets (so it's only `[parent_node, children]`) and having `Sidekick.Supervisor.start_link/2` instead?

sasa1977 commented on lib/sidekick/sidekick_supervisor.ex:
```diff
@@ -0,0 +1,41 @@
+    Parent.GenServer.start_link(__MODULE__, [parent_node, children], name: __MODULE__)
+  end
+
+  @impl true
```
```suggestion
  @impl GenServer
```

sasa1977 commented on lib/sidekick/sidekick_supervisor.ex:
```diff
@@ -0,0 +1,41 @@
+    end
+  end
+
+  @impl true
```
```suggestion
  @impl GenServer
```

sasa1977 commented on lib/sidekick/sidekick_supervisor.ex:
```diff
@@ -0,0 +1,41 @@
+
+  def handle_info({:EXIT, _pid, _reason}, state), do: {:noreply, state}
+
+  @impl true
```
```suggestion
  @impl GenServer
```

sasa1977 commented on lib/sidekick/sidekick_supervisor.ex:
```diff
@@ -0,0 +1,41 @@
+
+    all_started =
+      Enum.map(children, fn spec -> Parent.start_child(spec) end)
+      |> Enum.all?(&({:ok, _pid} = &1))
```
Could use [Parent.start_all_children](https://hexdocs.pm/parent/Parent.html#start_all_children!/1) instead. This function will raise, but I think that's fine.

sasa1977 commented on lib/sidekick/sidekick_supervisor.ex:
```diff
@@ -0,0 +1,41 @@
+    Logger.debug("Termination #{__MODULE__}")
+    Parent.shutdown_all()
+    :init.stop()
+    state
```
No need to return `state` from terminate.

sasa1977 commented on lib/sidekick/sidekick_supervisor.ex:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  @impl true
+  @spec init([...]) :: {:ok, nil} | {:stop, {:shutdown, :could_not_start_all_children}}
```
No need to spec callbacks, since those are never called directly.

sasa1977 commented on mix.exs:
```diff
@@ -14,10 +14,41 @@ defmodule Ci.MixProject do
+
+  defp get_version(app) do
+    path = :code.lib_dir(app)
+    {:ok, [{:application, ^app, properties}]} = :file.consult(Path.join(path, "ebin/#{app}.app"))
```
```suggestion
    {:ok, [{:application, ^app, properties}]} = :file.consult(Path.join([path, ebin, "#{app}.app"]))
```

sasa1977 commented on lib/sidekick/sidekick.ex:
```diff
@@ -0,0 +1,81 @@
+
+  defp start_node_command(sidekick_node, parent_node) do
+    {:ok, command} = :init.get_argument(:progname)
+    paths = Enum.join(:code.get_path(), " , ")
```
Why leading whitespace (before the comma)?

Hanspagh replied:
If I dont add this whitespace I get 
```
init terminating in do_boot ({undef,[{Elixir.Sidekick,start_sidekick,[[_]],[]},{init,start_em,1,[]},{init,do_boot,3,[]}]})
```
not entirely sure why this is the case

sasa1977 reviewed (commented):
Thanks, looks like a pretty good start to me :-)
Had a couple of questions, and various minor remarks/nitpicks, but nothing very dramatic. 

There are a few possible other changes that might be worth doing, but I really need to try this in practice to see if it makes sense, so I didn't want to elaborate them here.

I propose we address these minor remarks I mentioned, merge the PR, and then I can play with the code and report back.

Hanspagh committed:
review

sasa1977 committed:
Stop sidekick on failed connect

sasa1977 committed:
Remove needless spec

This is internal API, so we don't need the spec, which simplifies the implementation

sasa1977 committed:
Extract node starting code

sasa1977 committed:
Improve startup failure handling

sasa1977 committed:
Auto-start distributed node

Simplifies running if parent is not started in distributed mode.

sasa1977 committed:
Avoid passing parent node name

No need to make this parameterizable.

sasa1977 committed:
Refactor nodename building

sasa1977 committed:
Refactor startup code

sasa1977 committed:
Pass each path as a separate `-pa` arg

sasa1977 committed:
Improve naming

sasa1977 committed:
Add a basic test

sasa1977 committed:
Fix typespec

sasa1977 committed:
Make sure epmd is running

sasa1977 committed:
Use `System` where possible

sasa1977 committed:
Ensure required apps are started

sasa1977 committed:
Simplify startup flow

sasa1977 committed:
Simplify boot error handling

If the boot function crashes, the node will be taken down, so no need for manual stop.

sasa1977 committed:
Make ack more explicit

Node connects can happen implicitly, so instead we're relying on message passing. This also removes the need to invoke `monitor_nodes`.

sasa1977 committed:
Use correct name arg

sasa1977 committed:
Prevent early self-termination

sasa1977 committed:
Improve typespecs

Hanspagh commented:
hi @sasa1977 
Did you have a chance to play around with this :)

sasa1977 commented:
Hey,

I was planing to send you a mail these days. I actually played with this 
as soon as you submitted and made a PR on top of your PR, which you can 
find at https://github.com/Hanspagh/ci/pull/1
It appears you didn't get any notification?

Could you take a look at my proposed changes and let me know what you think?


On 31. 03. 2021. 00:29, Hans wrote:
>
> hi @sasa1977 <https://github.com/sasa1977>
> Did you have a chance to play around with this :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sasa1977/ci/pull/4#issuecomment-810621025>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AABROARAZ4N5B7HNMPEV7SLTGJGGNANCNFSM4ZGDG6SA>.
>

Hanspagh commented:
Ahh cool. I am sorry I missed this, I will have a look as soon as possible

Hanspagh committed:
force name on sidekick init

Hanspagh committed:
extend tests

Hanspagh commented:
Hi @sasa1977 
I have extended the test a bit, let me know what you think.
How do we avoid having the `Process.sleep()`?

sasa1977 commented on test/sidekick_test.exs:
```diff
@@ -2,6 +2,18 @@ defmodule SidekickTest do
+
+    # Shutting down the supervisor should close down the node
+    Supervisor.stop(pid)
+    Process.sleep(5000)
```
You could instead monitor the test node with [Node.monitor](https://hexdocs.pm/elixir/Node.html#monitor/2) and await the corresponding `:nodedown` message (using `assert_receive` with a larger timeout).

sasa1977 commented on test/sidekick_test.exs:
```diff
@@ -2,6 +2,18 @@ defmodule SidekickTest do
+    assert Node.list() == [:"test@127.0.0.1"]
+
+    # Supervisor should be started on the Node
+    pid = :rpc.call(:"test@127.0.0.1", Process, :whereis, [Sidekick.Supervisor])
```
I think that `GenServer.whereis({Sidekick.Supervisor, :"test@127.0.0.1"})` would also work.

sasa1977 reviewed (commented):
Another test that would be useful is checking that the sidekick node stops if the process that started it stops. Here's how it could be done:

1. From the test process start an agent
2. During the agent's initialization start a sidekick node
3. In the test process monitor the sidekick node
4. Stop the agent process
5. assert_receive `:nodedown` message

I'll take a look tomorrow to see if there are some more tests worth adding, but I think we're close to getting this merged.

Hanspagh commented:
> Another test that would be useful is checking that the sidekick node stops if the process that started it stops. Here's how it could be done:
> 
>     1. From the test process start an agent
> 
>     2. During the agent's initialization start a sidekick node
> 
>     3. In the test process monitor the sidekick node
> 
>     4. Stop the agent process
> 
>     5. assert_receive `:nodedown` message
> 
> 
> I'll take a look tomorrow to see if there are some more tests worth adding, but I think we're close to getting this merged.

I added two more tests.
Once that assert that sidekick does not shutdown even if the caller process gets killed
And another one asserting that if the parent **node** is killed then the sidekick is stopped, but here I have a problem with the following error
```
{:badrpc, {:EXIT, {{:case_clause, {:error, {{:already_started, #PID<20288.58.0>}, {:child, :undefined, :net_sup_dynamic, {:erl_distribution, :start_link, [[:"37s3sgdeeltf5aq42bnq2jb3za@127.0.0.1"], false, :net_sup_dynamic]}, :permanent, 1000, :supervisor, [:erl_distribution]}}}}, [{Sidekick, :ensure_distributed!, 0, [file: 'lib/sidekick/sidekick.ex', line: 18]}, {Sidekick, :start, 2, [file: 'lib/sidekick/sidekick.ex', line: 4]}]}}}
```
It looks like something goes wrong then we call `ensure_distributed!` in the parent node, but I can not figure out what it is. If I remove the `ensure_distributed!` the test parses

sasa1977 commented:
Oh, I'm sorry, I completely forgot we currently don't link the node to the starter process. This is something that needs to be implemented in the next step (as mentioned in [this comment](https://github.com/Hanspagh/ci/pull/1#issuecomment-801191822)). So you can drop the 2nd test, and even the first one was not needed because we'll change that behaviour next. We can then merge this, and proceed to link the node to the starter process.

Hanspagh commented:
I forgot we wanted to add the process link aswell. I will revert the test and get stated on the next steps

sasa1977 commented:
Cool. I merged this, so in the next PR let's switch to process link. I think this will simplify some other things (we might not need to monitor the parent node), and allow us to expand the tests.

sasa1977 commented:
The idea is to link `Sidekick.Supervisor` to the starter process and handle corresponding exit message by stopping the supervisor (which should then stop the node due to the logic in terminate). Link should also make sure that the starter process gets the exit signal if for some reasons the sidekick node stops, which I think makes sense (but I'm open to counter arguments). 

With that mechanism in place, we can remove `Node.monitor` and nodedown handling from the sidekick supervisor.

Hanspagh commented:
I started on this yesterday, but it seems process.link is not the way to
go, is it because the process is on an other node? I will give it a try
with monitors today.

On Thu, 8 Apr 2021, 23.34 Saša Jurić, ***@***.***> wrote:

> The idea is to link Sidekick.Supervisor to the starter process and handle
> corresponding exit message by stopping the supervisor (which should then
> stop the node due to the logic in terminate). Link should also make sure
> that the starter process gets the exit signal if for some reasons the
> sidekick node stops, which I think makes sense (but I'm open to counter
> arguments).
>
> With that mechanism in place, we can remove Node.monitor and nodedown
> handling from the sidekick supervisor.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sasa1977/ci/pull/4#issuecomment-816227608>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH2DIC2S7BIPFZJX3GV7DLTHYOPZANCNFSM4ZGDG6SA>
> .
>

sasa1977 commented:
> but it seems process.link is not the way to go, is it because the process is on an other node?

It should work. Here's a quick local test:

```
iex --name foo@127.0.0.1
iex --name bar@127.0.0.1

# on bar:
iex> Process.register(self(), :bar)

# on foo:
iex> pid = :rpc.call(:"bar@127.0.0.1", Process, :whereis, [:bar])
iex> Process.link(pid)
iex> Process.flag(:trap_exit, true)

# kill bar with ctrl-c ctrl-c

# on foo:
iex> flush
{:EXIT, pid, :noconnection}
```

Which problems did you experience?

Hanspagh commented:
I had a problem getting the test to parse when creating an agent and
starting the sidekick from there.
I never received the nodedown signal, I will give it an other shot later
today.

On Fri, 9 Apr 2021 at 09:44, Saša Jurić ***@***.***> wrote:

> but it seems process.link is not the way to go, is it because the process
> is on an other node?
>
> It should work. Here's a quick local test:
>
> iex --name ***@***.***
> iex --name ***@***.***
>
> # on bar:
> iex> Process.register(self(), :bar)
>
> # on foo:
> iex> pid = ***@***.***", Process, :whereis, [:bar])
> iex> Process.link(pid)
> iex> Process.flag(:trap_exit, true)
>
> # kill bar with ctrl-c ctrl-c
>
> # on foo:
> iex> flush
> {:EXIT, pid, :noconnection}
>
> Which problems did you experience?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sasa1977/ci/pull/4#issuecomment-816486533>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH2DIDGZGKUV2QWRSXFUALTH2V6LANCNFSM4ZGDG6SA>
> .
>


-- 
Med venlig hilsen / Best Regards
Hans Kristian Pagh

sasa1977 commented:
If you need help, push some temp branch and tell me how to reproduce the problem.

Note that with links the sidekick node doesn't need to (in fact shouldn't) monitor the parent node.
